### PR TITLE
multistage docker build

### PR DESCRIPTION
**Pull request by wallies**

### CODE REVIEW

**Comment by ishaan-jaff** [comment]
@wallies what is the benefit of doing it like this ?

**Comment by ghost** [comment]
closing due to inactivity.

**Comment by wallies** [comment]
@krrishdholakia Closed due to inactivity after a day? Very confusing, most of the other PRs have had no activity after weeks. Most people in opensource take more than a day to respond as they have day jobs. If you just said we are working on this, thats fine as well, inactivity makes no sense.

To answer your question, the benefit of doing this removes things like python-dev out of the container and I can run a much smaller container if i need for runtime e.g. i can use chainguard images https://edu.chainguard.dev/chainguard/chainguard-images/reference/python/. Multi-stage builds is also the generally accepted way to build images, as I said earlier allows much smaller and more secure images for runtime and separating build and runtime is much nicer.

So i built the docker image from main its size is 903.8 MiB
Here is the vulnerability scan from trivy

ghcr.io/berriai/litellm:main-v1.10.3 (debian 12.2)
==================================================
Total: 332 (UNKNOWN: 0, LOW: 226, MEDIUM: 71, HIGH: 33, CRITICAL: 2)

If i do the same with this PR. 
The size is 582MB

**Comment by ishaan-jaff** [comment]
sorry about that @wallies, I re-opened your PR. Thanks for the response and your contribution. We appreciate it

File: Dockerfile [hunk 1]
```diff
@@ -13,16 +16,34 @@ RUN apt-get update && \
...
+# Make a virtualenv that we can copy across multistage builds
+ENV VIRTUAL_ENV=/venv
+RUN python -m venv $VIRTUAL_ENV
+# "Activate" the virtualenv
+ENV PATH="$VIRTUAL_ENV/bin:$PATH"
 
 # Install any needed packages specified in requirements.txt
-RUN pip wheel --no-cache-dir --wheel-dir=wheels -r requirements.txt
-RUN pip install --no-cache-dir --find-links=wheels -r requirements.txt
+RUN pip install --no-cache-dir -r /app/requirements.txt
```
Comment (ishaan-jaff) [inline Dockerfile]: can you not remove `pip wheel` ? It speeds up installing requirements.txt 

cc @PSU3D0

**Review by ishaan-jaff — CHANGES_REQUESTED** [review]

**Comment by ishaan-jaff** [comment]
left a review @wallies

File: Dockerfile [hunk 2]
```diff
@@ -13,16 +16,34 @@ RUN apt-get update && \
...
 
 # Copy the current directory contents into the container at /app
-COPY . /app
+COPY requirements.txt .
+
+# Make a virtualenv that we can copy across multistage builds
+ENV VIRTUAL_ENV=/venv
+RUN python -m venv $VIRTUAL_ENV
+# "Activate" the virtualenv
+ENV PATH="$VIRTUAL_ENV/bin:$PATH"
```
Comment (PSU3D0) [inline Dockerfile]: Why not use the system environment with `pip wheel`?

File: Dockerfile [hunk 3]
```diff
@@ -1,8 +1,11 @@
 # Base image
-ARG LITELLM_BASE_IMAGE=python:3.9-slim
+ARG LITELLM_BASE_IMAGE=python:3.9
```
Comment (PSU3D0) [inline Dockerfile]: As we are separating build and runtime images, can you rename this to `LITELLM_BUILD_IMAGE` to be more explicit?

**Review by PSU3D0 — CHANGES_REQUESTED** [review]

File: Dockerfile [hunk 4]
```diff
@@ -1,8 +1,11 @@
...
 
-# allow users to specify, else use python 3.9-slim
-FROM $LITELLM_BASE_IMAGE
+# Runtime image
+ARG LITELLM_RUNTIME_IMAGE=python:3.9-slim
```
Comment (PSU3D0) [inline Dockerfile]: Come to think of it, can we get away with `alpine` as the default? This should reduce build size further

@krrishdholakia Foresee any dependencies that might not play well with ARM?

File: Dockerfile (hunk 2 above)
Comment (wallies) [reply Dockerfile]: you could do a pip install and then copy the packages over from the lib directory like chainguard does https://edu.chainguard.dev/chainguard/chainguard-images/reference/python/

File: Dockerfile (hunk 1 above)
Comment (wallies) [reply Dockerfile]: yes it speeds it up, but its just double handling, pip wheel is pulling the packages down then installing them the pip install is doing that in one step.

File: Dockerfile (hunk 4 above)
Comment (wallies) [reply Dockerfile]: yeh alpine or wolfi base https://edu.chainguard.dev/open-source/wolfi/wolfi-with-dockerfiles/.

File: Dockerfile (hunk 2 above)
Comment (PSU3D0) [reply Dockerfile]: A virtualenv is not necessary in this case. We'll never have multiple environments in this docker container. A system-wide install is better and results in less unneccesary files (due to the virtualenv).

The following sample is a very explicit and minimal way to install from wheel-directory that minimizes container size.

```
COPY --from=python-build-stage /usr/src/app/wheels  /wheels/

RUN pip install --no-cache-dir --no-index --find-links=/wheels/ /wheels/* \
  && rm -rf /wheels/
```

**Comment by wallies** [comment]
@krrishdholakia @PSU3D0 @ishaan-jaff i hope my comment didnt come off as negative. I just didnt understand what was happening. I opened this PR more as a discussion starter.

**Comment by ghost** [comment]
@wallies no worries - realized we have a bunch of open PRs and was doing cleanup. Appreciate your contribution!

**Comment by ghost** [comment]
what're next steps on this pr? @wallies @ishaan-jaff @PSU3D0

**Comment by ghost** [comment]
@wallies ?

**Comment by wallies** [comment]
@krrishdholakia should be good to go. I think we can make any alpine base image change later as that's a bugger change

**Comment by PSU3D0** [comment]
Looks good

**Comment by ghost** [comment]
@wallies are there any documentation changes required - https://docs.litellm.ai/docs/proxy/deploy#run-the-docker-image

**Comment by wallies** [comment]
> @wallies are there any documentation changes required - https://docs.litellm.ai/docs/proxy/deploy#run-the-docker-image

I updated docs

**Comment by ghost** [comment]
Great work!

**Pull request merged**